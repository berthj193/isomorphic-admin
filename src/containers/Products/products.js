import React, { Component } from 'react';
import PropTypes from 'prop-types';
import autoBind from 'react-autobind';
import columns from './columns';
import { Icon, Tooltip } from 'antd';
import ItemList from '../../components/list/item-list';
import productQueries from '../../graphql/products';
import withStoresManager from '../../components/apolloProviders/withStoresManager';
import DeleteButton from '../../components/uielements/deleteButton';

export class Products extends Component {
  static propTypes = {
    products: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      default_name: PropTypes.string,
      default_description: PropTypes.string,
      default_master_sku: PropTypes.string,
      stores: PropTypes.arrayOf(PropTypes.object),
    })),
    error: PropTypes.string,
    history: PropTypes.shape({
      push: PropTypes.func,
    }),
    mutate: PropTypes.func,
  };

  static defaultProps = {
    products: [],
  };

  constructor() {
    super();
    autoBind(this);
  }

  dataSource() {
    return this.props.products.map(product => ({
      key: product.id,
      id: product.id,
      name: product.default_name,
      description: product.default_description,
      sku: product.default_master_sku,
      stores_count: product.stores.length,
    }));
  }

  editProductVariants = id => e => {
    e.preventDefault();
    e.stopPropagation();
    this.props.history.push(`/dashboard/product_variants/${id}`);
  };

  deleteProduct = id => e => {
    e.preventDefault();
    e.stopPropagation();
    this.props.mutate({
      mutation: productQueries.destroyProduct,
      variables: {
        id,
      },
      messages: ['Product has been removed successfully'],
    });
  };

  actionColumn = ({ id }) => (
    <ul className="table-action">
      <li>
        <Tooltip title="Edit product variants">
          <a onClick={this.editProductVariants(id)}><Icon type="tags-o" /></a>
        </Tooltip>
        <DeleteButton
          onConfirm={this.deleteProduct(id)}
          tooltipText="Remove a product"
        />
      </li>
    </ul>
  );

  columns = [
    ...columns,
    {
      title: 'Action',
      dataIndex: '',
      key: 'a',
      render: this.actionColumn,
    },
  ];

  render() {
    return (
      <ItemList
        data={this.dataSource()}
        error={this.props.error}
        title="Products"
        columns={this.columns}
        buttons={[{
          title: 'New product',
          link: '/dashboard/products/new',
        }]}
      />
    );
  }
}

export default withStoresManager(Products,
  {
    query: productQueries.getProducts(),
  }
);
