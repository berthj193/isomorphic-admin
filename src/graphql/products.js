import gql from 'graphql-tag';

export const getListQuery = () => `
query {
  products {
    id
    default_name
    default_description
    default_master_sku
    stores {
      key
    }
    store_products {
      master_sku
    }
    product_variants {
      default_sku
    }
  }
}`;

export const getDetailsQuery = id => `
query {
  products( filter: {
    id: ${id}
  }) {
    default_name
    default_description
    default_master_sku
    stores {
      key
    }
    store_products {
      master_sku
    }
    product_variants {
      default_sku
    }
  }
}`;

export const getAutocompleteQuery = `
query {
  products {
    id
    default_name
  }
}`;

export const getCreateQuery = () => `
  mutation createProduct(
    $default_name: String
    $default_description: String
    $default_master_sku: String!
    $store_keys: [String]
  ) {
    createProduct(
      default_name: $default_name
      default_description: $default_description
      default_master_sku: $default_master_sku
      store_keys: $store_keys
    )
    {
      id
      default_name
      default_description
      default_master_sku
      stores {
        key
      }
      store_products {
        master_sku
      }
      product_variants {
        default_sku
      }
    }
  }
`;

export const getUpdateQuery = () => `
mutation updateProduct(
  $id: ID!
  $default_name: String
  $default_description: String
  $default_master_sku: String!
) {
  updateProduct(
    id: $id
    default_name: $default_name
    default_description: $default_description
    default_master_sku: $default_master_sku
  )
  {
    id
    default_name
    default_description
    default_master_sku
    stores {
      key
    }
    store_products {
      master_sku
    }
    product_variants {
      default_sku
    }
  }
}`;

export const destroyProductQuery = `
mutation destroyProduct(
  $id:ID!
) {
  destroyProduct(id:$id)
}`;

export default {
  getProducts: () => gql(getListQuery()),
  getProduct: id => gql(getDetailsQuery(id)),
  createProduct: () => gql(getCreateQuery()),
  updateProduct: () => gql(getUpdateQuery()),
  productsAutocomplete: gql(getAutocompleteQuery),
  destroyProduct: gql(destroyProductQuery),
};
