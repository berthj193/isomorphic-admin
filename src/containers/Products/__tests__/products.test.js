import React from 'react';
import { Products } from '../products';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<Products />', () => {
  const defaultProps = {};

  const getComponent = (props = {}) => shallow(<Products {...Object.assign(defaultProps, props)} />);

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render properly', () => {
    expect(toJson(getComponent())).toMatchSnapshot();
  });

  it('should map products', () => {
    const mockStores = [
      {
        id: 1,
        default_name: 'MyLittlePony Toy',
        default_description: 'Representation of unicorn',
        default_master_sku: 'r41nb0w',
        stores: [{}, {}],
      },
      {
        id: 2,
        default_name: 'Necronomicon',
        default_description: 'Ph\'nglui mglw\'nafh Cthulhu R\'lyeh wgah\'nagl fhtagn',
        default_master_sku: 'd3v1l',
        stores: [],
      },
    ];
    const component = getComponent({ products: mockStores });

    expect(component.instance().dataSource()).toMatchSnapshot();
  });
});

