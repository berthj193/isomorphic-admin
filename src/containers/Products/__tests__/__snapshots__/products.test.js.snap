// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Products /> should map products 1`] = `
Array [
  Object {
    "description": "Representation of unicorn",
    "id": 1,
    "key": 1,
    "name": "MyLittlePony Toy",
    "sku": "r41nb0w",
    "stores_count": 2,
  },
  Object {
    "description": "Ph'nglui mglw'nafh Cthulhu R'lyeh wgah'nagl fhtagn",
    "id": 2,
    "key": 2,
    "name": "Necronomicon",
    "sku": "d3v1l",
    "stores_count": 0,
  },
]
`;

exports[`<Products /> should render properly 1`] = `
<withRouter(ItemList)
  buttons={
    Array [
      Object {
        "link": "/dashboard/products/new",
        "title": "New product",
      },
    ]
  }
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "Name",
      },
      Object {
        "dataIndex": "description",
        "key": "description",
        "title": "Description",
      },
      Object {
        "dataIndex": "sku",
        "key": "sku",
        "title": "SKU",
      },
      Object {
        "dataIndex": "stores_count",
        "key": "stores_count",
        "title": "No. Stores",
      },
      Object {
        "dataIndex": "",
        "key": "a",
        "render": [Function],
        "title": "Action",
      },
    ]
  }
  data={Array []}
  title="Products"
/>
`;
